Login: Handle Jetpack site connection for self-hosted sites

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #7846 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
When working on #7597 and #7745, we added Jetpack connection handling for accounts without connection to Jetpack sites. 

There's another case where we need to handle Jetpack connection - when a site has not registered with Jetpack. This is when we need to handle site connection before proceeding to account connection.

To successfully register Jetpack connection for a site, we need to request the connection URL, then access it with the same cookies used to request the URL. Otherwise, the URL will be considered expired.

The solution introduced in this PR includes:
- Only show Jetpack error if a user tries to log in to a site that does not have Jetpack installed or activated. If the site only misses Jetpack connection, show them the account mismatch screen, where we can help them register the Jetpack connection on their site.
- Retrieve cookies from WordPressOrgNetwork
- Send retrieved cookies to AuthenticatedWebViewController
- Add a link to show terms and conditions to be agreed on when connecting Jetpack.
- Also: update the primary button title on the Account mismatch screen to "Connect Jetpack to your account" to align with Android.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Handling Jetpack connection when logging in with site credentials:
- Pre-requisite: create a JN site with Jetpack installed, activated but not connected. The site may or may not have Woo.
- Log out of the app if needed. On the prologue screen, select Enter site address.
- Enter the address of the JN site.
- Select Log in with site credentials.
- Enter the credentials and tap Continue. Notice that the account mismatch screen is then displayed.
- Notice a new text at the bottom of the screen regarding the terms to be agreed upon when connecting Jetpack.
- Tap Connect Jetpack to your account. Notice that a web view is displayed navigating to the Jetpack authorization screen. Tap Authorize, notice that the app will then navigate to the WP.com login flow and show the home screen after authentication succeeds.

Handling Jetpack connection when logging in with a WP.com account:
- Pre-requisite: create a JN site with Jetpack installed, activated but not connected. The site may or may not have Woo.
- Log out of the app if needed. On the prologue screen, select Enter site address.
- Enter the address of the JN site.
- Enter WP.com credentials and tap Continue. Notice that the account mismatch screen is then displayed.
- Notice a new text at the bottom of the screen regarding the terms to be agreed upon when connecting Jetpack.
- Tap Connect Jetpack to your account. Notice that the site credential login flow is presented. Enter the credentials and tap Continue.
- Notice that a web view is displayed navigating to the Jetpack authorization screen. Tap Authorize, notice that the app will then navigate to the WP.com login flow and show the home screen after authentication succeeds.

Please feel free to test with a mismatch WP.com site. Notice that there is no terms text at the bottom of the account mismatch screen.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/5533851/195795881-12fe325a-fd53-4d37-a910-45837241d9e8.mp4




---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Navigate to account mismatch if the site already has Jetpack installed and activated

itsmeichigo committed:
Remove the check for the site address prefix for jetpack connection completion

itsmeichigo committed:
Add extension to retrieve cookie from WordPressOrgNetwork

itsmeichigo committed:
Add new a parameter for cookies in AuthenticatedWebViewController

itsmeichigo committed:
Send cookie from WordPressOrgNetwork to AuthenticatedWebViewController for Jetpack site connection

itsmeichigo committed:
Update title for the primary button to match Android's

itsmeichigo committed:
Check for jetpack connection page to not detect completion

itsmeichigo committed:
Fix unit test failures

itsmeichigo committed:
Add a new helper method to NSMutableAttributedString to set a link to a substring

itsmeichigo committed:
Add termsLabelText to ULAccountMismatchViewModel

itsmeichigo committed:
Add a text view for the terms label in ULAccountMismatchViewController

itsmeichigo committed:
Only show terms label if the site is self-hosted

itsmeichigo committed:
Don't authenticate user with WP.com if cookies are sent to AuthenticatedWebViewController

itsmeichigo committed:
Only authenticate for WP.com automatically if no cookies are provided

itsmeichigo committed:
Fix test build failure

itsmeichigo committed:
Add comment for the shared cookies

itsmeichigo committed:
Add comments for the check of self-hosted site when constructing terms text

selanthiraiyan commented on WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift:
```diff
@@ -128,9 +133,14 @@ private extension AuthenticatedWebViewController {
         } else {
+            for cookie in cookies {
+                webView.configuration.websiteDataStore.httpCookieStore.setCookie(cookie)
+            }
```
❓  Is this for automatically logging into the site using .org credentials? 

Would it make sense to use a helper method like `func setCookiesToAuthenticateForDotOrgSignIn()` to explain this? (Sorry, I can't come up with a better name. 😄 )

selanthiraiyan commented on WooCommerce/Classes/Authentication/Navigation Exceptions/WrongAccountErrorViewModel.swift:
```diff
@@ -117,6 +123,32 @@ final class WrongAccountErrorViewModel: ULAccountMismatchViewModel {
+        $isSelfHostedSite
+            .map { [weak self] isSelfHosted -> NSMutableAttributedString in
+                // only shows terms text if the site is self-hosted,
+                // since the user cannot handle Jetpack connection themselves on WP.com sites.
```
Thanks for adding comments. 👏

selanthiraiyan commented on WooCommerce/Classes/Authentication/Navigation Exceptions/WrongAccountErrorViewModel.swift:
```diff
@@ -117,6 +123,32 @@ final class WrongAccountErrorViewModel: ULAccountMismatchViewModel {
                 self?.analytics.track(event: .LoginJetpackConnection.jetpackConnectionErrorShown(selfHostedSite: isSelfHosted))
             }
 
+        $isSelfHostedSite
```
Super nit: Would it make sense to move this out of `viewDidLoad` to reduce the number of lines in `viewDidLoad`?

selanthiraiyan reviewed (approved):
**Testing**
It works as expected ✅ 

**Code**
Left a few non-blocking comments/questions. 

LGTM 🚢 🎉

itsmeichigo committed:
Move screen view tracking and terms text configuration to helper methods

selanthiraiyan commented:
>Please feel free to test with a mismatch WP.com site. Notice that there is no terms text at the bottom of the account mismatch screen.

I tested this for a self-hosted site (`https://genuine-barnacle.jurassic.ninja/`) and a WPCOM account which isn't associated with that site. 

i.e. `https://public-api.wordpress.com/rest/v1.1/me/sites` response for that WPCOM account doesn't have `https://genuine-barnacle.jurassic.ninja/` in it. 

But I still see the terms text at the bottom. Am I missing something here?

<img src="https://user-images.githubusercontent.com/524475/196100121-8ef7bfc8-2bbc-4f86-b993-3563bf83c98e.png" width="300"/>

itsmeichigo committed:
Make `Remote` subclass NSObject for Obj-C delegate conformance

itsmeichigo committed:
Add new readonly variable `session` for Network and remove cookies from `WordPressOrgNetwork`

itsmeichigo committed:
Handle Jetpack site connection in JetpackConnectionRemote

itsmeichigo committed:
Update `JetpackConnectionStore` to handle Jetpack site connection

itsmeichigo committed:
Remove cookie injection from AuthenticatedWebViewController

itsmeichigo committed:
Remove redundant white space

itsmeichigo committed:
Merge branch 'trunk' into try/jetpack-site-connection

itsmeichigo commented:
Thank you @selanthiraiyan for the review!

> But I still see the terms text at the bottom. Am I missing something here?

This is intentional. So the expectation is that the terms text is only available for self-hosted sites because we can help users connect Jetpack on their site - and not for WP.com sites because they need an invitation from a site admin to be connected.

So to test that case about the mismatch WP.com site, you need to log in to a WP.com (Atomic) site with a mismatch WP.com account instead. The terms text should not be available then.

**Another update**: I've added a few more commits to restore the automatic WP.com authentication in `AuthenticatedWebViewController` when handling Jetpack site connection.

- Previously, I retrieved the cookies from `WordPressOrgNetwork` and injected them into the `AuthenticatedWebViewController`:   
  - The cookies make it possible to handle the site connection URL
  - However, they conflict with the automatic WP.com authentication, so I added a workaround to remove the authentication if there are injected cookies. 
  - This means that users will have to enter their WP.com credentials again in the web view even if they already did in the login flow earlier.
 
- Then, I followed Hicham's suggestion [p1665912798859479/1665502423.023859-slack-C03L1NF1EA3]:
  - Removed the injection of cookies to  `AuthenticatedWebViewController`.
  - Added a read-only variable `session` to `Network`. This is needed to retrieve cookies from the session of the network client.
  - Updated `JetpackConnectionNetwork` to handle Jetpack site connection registration. I used `URLSession` for simplicity to handle redirection; the idea is to inject the necessary cookies and request the site connection URL. 
  - When the site connection URL is triggered, a redirection will be requested to navigate to the account connection URL. We want to show this URL in a web view, so the redirection should be canceled (using `URLSessionDataDelegate`). 
  - The redirected URL will then be saved and sent to the `JetpackConnectionStore`. 
  - A web view will then be displayed and show the account connection URL, with the WP.com authentication already handled. If the WP.com account has 2FA enabled, a code may still be requested (see the recording below).


https://user-images.githubusercontent.com/5533851/196150289-9ed046a9-a8d0-4208-ad0f-600ab9118c84.mp4

I know this is all very complicated 😅 Please feel free to ping me with any questions you have, I'll try my best to explain.

I haven't added new unit tests for the new updates; I'm not sure what I can do about that. Please also leave any suggestions if necessary.

selanthiraiyan commented:
>This is intentional. So the expectation is that the terms text is only available for self-hosted sites because we can help users connect Jetpack on their site - and not for WP.com sites because they need an invitation from a site admin to be connected.

I understand. Thanks for explaining this.

selanthiraiyan commented:
Thanks for the detailed explanation about the new updates, @itsmeichigo 

This still LGTM. 🎉
